File: haskell/022/Main.hs
{-
    ProjectEuler 022
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc Main.hs -o bin/main`
-}

module Main where

import Data.Char
import Data.List

main :: IO ()
main = do
    contents <- readFile "./names.txt"
    let names  = sort . read $ "[" ++ contents ++ "]"
        scores = zipWith (*) (map worth names) [1 ..]
    print $ foldl' (+) 0 scores
    where
        worth name = foldl' (+) 0 $ map (\c -> ord c - ord 'A' + 1) name

File: ocaml/022/main.ml
(*
    ProjectEuler 022
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main -linkpkg -package core -thread main.ml`
*)

open Core

let value_of_char c = Char.code c - Char.code 'A' + 1

let value_of_name name =
    let chars = Core_string.to_list name in
    List.fold_left (+) 0 (List.map value_of_char chars)

let main _ =
    let ic = open_in "./names.txt" in
    try
        let names_string = input_line ic in
        let names_quoted = Core_string.split names_string ',' in
        let names        = List.map (fun s -> Core_string.slice s 1 (-1)) names_quoted in
        let names_sorted = List.sort String.compare names in
        let scores       = List.mapi (fun i s -> (i + 1) * value_of_name s) names_sorted in
        let total_score  = List.fold_left (+) 0 scores in
        print_int total_score;
        print_newline ()
    with
    | e ->
        close_in_noerr ic;
        raise e

let _ = main ()
